plugins/functions/string.c

```diff
@@ -88,24 +94,77 @@ func_string_substring(grn_ctx *ctx, int n_args, grn_obj **args,
   grn_obj *target;
   grn_obj *from_raw;
   grn_obj *length_raw = NULL;
+  grn_obj *default_value = NULL;
+  grn_obj *options = NULL;
   int64_t from = 0;
   int64_t length = -1;
   const char *start = NULL;
   const char *end = NULL;
   grn_obj *substring;
 
-  if (!(n_args == 2 || n_args == 3)) {
+  if (n_args < 2 || n_args > 4) {
     GRN_PLUGIN_ERROR(ctx, GRN_INVALID_ARGUMENT,
                      "[string_substring] "
-                     "wrong number of arguments (%d for 2..3)",
+                     "wrong number of arguments (%d for 2..4)",
                      n_args);
     return NULL;
   }
 
   target = args[0];
   from_raw = args[1];
-  if (n_args == 3) {
-    length_raw = args[2];
+
+  if (n_args >= 3) {
+    grn_obj *length_or_options = args[2];
+    if (grn_obj_is_number_family_bulk(ctx, length_or_options)) {
+      length_raw = length_or_options;
+    } else if (length_or_options->header.type == GRN_TABLE_HASH_KEY) {
+      options = length_or_options;
+    } else {
+      grn_obj inspected;
+
+      GRN_TEXT_INIT(&inspected, 0);
+      grn_inspect(ctx, &inspected, length_or_options);
+      GRN_PLUGIN_ERROR(ctx, GRN_INVALID_ARGUMENT,
+                       "%s "
+                       "3rd argument must be long or hash table: %.*s",
+                       "[string_substring]",
```

Comment:
How about define `const char *tag = "[string_substring]"` at the head of this function and use it in this function?

Reply:
fixed it.
In some cases `#define` like `#define string_substring tag "[string_substring]"` is used instead of `const char* tag`  because it is useful for joining strings.

Reply:
It makes sense.